panic: runtime error: invalid memory address or nil pointer dereference

```
{"level":"warn","time":"2024-09-12T16:51:09.787253248Z","component":"internal-rpc-server","error":{"error":"leaked iterators: 8","kind":"*withstack.withStack","stack":null},"namespace":"xxxxx","peer":"192.168.3.222:33482","shard":2,"time":"2024-09-12T16:51:09.787296002Z","message":"Replicate failed: could not get follower controller"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x19e8a1d]

goroutine 8725 [running]:
github.com/cockroachdb/pebble/sstable.(*Reader).readBlock(0xc000a10708, {0x247b8a0, 0x3593c00}, {0x2?, 0x0?}, 0x0, {0x0, 0x0}, 0x0, 0x0)
	/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.0/sstable/reader.go:558 +0x2bd
github.com/cockroachdb/pebble/sstable.(*Reader).readRangeDel(0x0?, 0x0?)
	/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.0/sstable/reader.go:459 +0x3e
github.com/cockroachdb/pebble/sstable.(*Reader).NewRawRangeDelIter(0xc000a10708)
	/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.0/sstable/reader.go:400 +0x29
github.com/cockroachdb/pebble.(*tableCacheShard).newIters(0xc0003d6fd0, {0x247b8a0, 0x3593c00}, 0xc006b9e480, 0xc00a758530, {0x0, 0x0, 0xc00a757bb0, {0x0, 0x0}}, ...)
	/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.0/table_cache.go:486 +0x44e
github.com/cockroachdb/pebble.(*tableCacheContainer).newIters(0xc00a758678?, {0x247b8a0?, 0x3593c00?}, 0x0?, 0x0?, {0x0, 0x0, 0xc00a757bb0, {0x0, 0x0}})
	/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.0/table_cache.go:137 +0x6e
github.com/cockroachdb/pebble.(*levelIter).loadFile(0xc00a7584c8, 0xc006b9e480, 0x1)
	/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.0/level_iter.go:678 +0x2d1
github.com/cockroachdb/pebble.(*levelIter).SeekGE(0xc00a7584c8, {0xc0031758c0, 0x25, 0x60}, 0x0)
	/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.0/level_iter.go:735 +0x90
github.com/cockroachdb/pebble.(*mergingIter).seekGE(0xc00a757de0, {0xc0031758c0?, 0x0?, 0x0?}, 0x0?, 0x0)
	/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.0/merging_iter.go:1050 +0x16f
github.com/cockroachdb/pebble.(*mergingIter).SeekGE(0xc00a757de0, {0xc0031758c0?, 0xc00302f8d0?, 0x1a4e68b?}, 0x8?)
	/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.0/merging_iter.go:1112 +0x65
github.com/cockroachdb/pebble.(*Iterator).iterFirstWithinBounds(0xc00a757808)
	/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.0/iterator.go:2010 +0x4c
github.com/cockroachdb/pebble.(*Iterator).First(0xc00a757808)
	/go/pkg/mod/github.com/cockroachdb/pebble@v1.1.0/iterator.go:1610 +0xea
github.com/streamnative/oxia/server/kv.(*Pebble).RangeScan(0xc0018db800, {0xc006d3db00?, 0xc00302fac0?}, {0xc0002503f0?, 0x25?})
	/src/oxia/server/kv/kv_pebble.go:501 +0xdc
github.com/streamnative/oxia/server/kv.(*notificationsTracker).ReadNextNotifications(0xc00154dea0, {0x247b910?, 0xc003ff2be0?}, 0x25767)
	/src/oxia/server/kv/notifications_tracker.go:171 +0x123
github.com/streamnative/oxia/server/kv.(*db).ReadNextNotifications(0xc002f9d0b0?, {0x247b910?, 0xc003ff2be0?}, 0xfffffffffffffffc?)
	/src/oxia/server/kv/db.go:665 +0x25
github.com/streamnative/oxia/server.(*leaderController).iterateOverNotifications(0xc0018db600, {0x247b910, 0xc003ff2be0}, {0x2483740, 0xc0046ae330}, 0x16?)
	/src/oxia/server/leader_controller.go:920 +0xa5
github.com/streamnative/oxia/server.(*leaderController).dispatchNotifications(0xc0018db600, {0x247b910, 0xc003ff2be0}, 0xc00196cb40, {0x2483740, 0xc0046ae330})
	/src/oxia/server/leader_controller.go:914 +0x2f6
github.com/streamnative/oxia/server.(*leaderController).GetNotifications.func1()
	/src/oxia/server/leader_controller.go:847 +0x58
github.com/streamnative/oxia/common.DoWithLabels.func1({0x247b8d8?, 0xc0046e82d0?})
	/src/oxia/common/pprof.go:46 +0x13
runtime/pprof.Do({0x247b910?, 0xc003ff2be0?}, {{0xc00237f140?, 0x0?, 0x0?}}, 0xc002712f30)
	/usr/local/go/src/runtime/pprof/runtime.go:51 +0x9d
github.com/streamnative/oxia/common.DoWithLabels({0x247b910, 0xc003ff2be0}, 0xc0046e8240, 0xc00196cbc0)
	/src/oxia/common/pprof.go:42 +0x356
created by github.com/streamnative/oxia/server.(*leaderController).GetNotifications in goroutine 8724
	/src/oxia/server/leader_controller.go:839 +0x2e9
```

--- a/server/leader_controller.go
+++ b/server/leader_controller.go
@@ -102,17 +102,20 @@ type leaderController struct {
 	headOffsetGauge         metrics.Gauge
 	commitOffsetGauge       metrics.Gauge
 	followerAckOffsetGauges map[string]metrics.Gauge
+
+	notificationDispatchers map[int64]*notificationDispatcher
 }
 
 func NewLeaderController(config Config, namespace string, shardId int64, rpcClient ReplicationRpcProvider, walFactory wal.Factory, kvFactory kv.Factory) (LeaderController, error) {
 	labels := metrics.LabelsForShard(namespace, shardId)
 	lc := &leaderController{
-		status:           proto.ServingStatus_NOT_MEMBER,
-		namespace:        namespace,
-		shardId:          shardId,
-		quorumAckTracker: nil,
-		rpcClient:        rpcClient,
-		followers:        make(map[string]FollowerCursor),
+		status:                  proto.ServingStatus_NOT_MEMBER,
+		namespace:               namespace,
+		shardId:                 shardId,
+		quorumAckTracker:        nil,
+		rpcClient:               rpcClient,
+		followers:               make(map[string]FollowerCursor),
+		notificationDispatchers: make(map[int64]*notificationDispatcher),
 
 		writeLatencyHisto: metrics.NewLatencyHistogram("oxia_server_leader_write_latency",
 			"Latency for write operations in the leader", labels),
@@ -963,110 +966,7 @@ func (lc *leaderController) appendToWalStreamRequest(request *proto.WriteRequest
 // ////
 
 func (lc *leaderController) GetNotifications(req *proto.NotificationsRequest, stream proto.OxiaClient_GetNotificationsServer) error {
-	// Create a context for handling this stream
-	ctx, cancel := context.WithCancel(stream.Context())
-
-	go common.DoWithLabels(
-		ctx,
-		map[string]string{
-			"oxia":  "dispatch-notifications",
-			"shard": fmt.Sprintf("%d", lc.shardId),
-			"peer":  common.GetPeer(stream.Context()),
-		},
-		func() {
-			if err := lc.dispatchNotifications(ctx, req, stream); err != nil && !errors.Is(err, context.Canceled) {
-				lc.log.Warn(
-					"Failed to dispatch notifications",
-					slog.Any("error", err),
-					slog.String("peer", common.GetPeer(stream.Context())),
-				)
-				cancel()
-			}
-		},
-	)
-
-	select {
-	case <-lc.ctx.Done():
-		// Leader is getting closed
-		cancel()
-		return lc.ctx.Err()
-
-	case <-ctx.Done():
-		return ctx.Err()
-
-	case <-stream.Context().Done():
-		// The stream is getting closed
-		cancel()
-		return stream.Context().Err()
-	}
-}
-
-func (lc *leaderController) dispatchNotifications(ctx context.Context, req *proto.NotificationsRequest, stream proto.OxiaClient_GetNotificationsServer) error {
-	lc.log.Debug(
-		"Dispatch notifications",
-		slog.Any("start-offset-exclusive", req.StartOffsetExclusive),
-	)
-
-	var offsetInclusive int64
-	if req.StartOffsetExclusive != nil {
-		offsetInclusive = *req.StartOffsetExclusive + 1
-	} else {
-		lc.Lock()
-		qat := lc.quorumAckTracker
-		lc.Unlock()
-
-		if qat == nil {
-			return errors.New("leader is not yet ready")
-		}
-		commitOffset := qat.CommitOffset()
-
-		// The client is creating a new notification stream and wants to receive the notification from the next
-		// entry that will be written.
-		// In order to ensure the client will positioned on a given offset, we need to send a first "dummy"
-		// notification. The client will wait for this first notification before making the notification
-		// channel available to the application
-		lc.log.Debug(
-			"Sending first dummy notification",
-			slog.Int64("commit-offset", commitOffset),
-		)
-		if err := stream.Send(&proto.NotificationBatch{
-			Shard:         lc.shardId,
-			Offset:        commitOffset,
-			Timestamp:     0,
-			Notifications: nil,
-		}); err != nil {
-			return err
-		}
-
-		offsetInclusive = commitOffset + 1
-	}
-
-	return lc.iterateOverNotifications(ctx, stream, offsetInclusive)
-}
-
-func (lc *leaderController) iterateOverNotifications(ctx context.Context, stream proto.OxiaClient_GetNotificationsServer, startOffsetInclusive int64) error {
-	offsetInclusive := startOffsetInclusive
-	for ctx.Err() == nil {
-		notifications, err := lc.db.ReadNextNotifications(ctx, offsetInclusive)
-		if err != nil {
-			return err
-		}
-
-		lc.log.Debug(
-			"Got a new list of notification batches",
-			slog.Int("list-size", len(notifications)),
-		)
-
-		for _, n := range notifications {
-			if err := stream.Send(n); err != nil {
-				return err
-			}
-		}
-
-		offsetInclusive += int64(len(notifications))
-	}
-
-	return ctx.Err()
+	return startNotificationDispatcher(lc, req, stream)
 }
 
 func (lc *leaderController) isClosed() bool {
@@ -1097,6 +997,10 @@ func (lc *leaderController) close() error {
 	lc.followerAckOffsetGauges = map[string]metrics.Gauge{}
 
 	err = lc.sessionManager.Close()
+
+	for _, nd := range lc.notificationDispatchers {
+		nd.close()
+	}
 
 	if lc.wal != nil {
 		err = multierr.Append(err, lc.wal.Close())

--- a/server/notifications_dispatcher.go
+++ b/server/notifications_dispatcher.go
@@ -1,0 +1,175 @@
+// Copyright 2023 StreamNative, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"sync/atomic"
+
+	"github.com/pkg/errors"
+
+	"github.com/streamnative/oxia/common"
+	"github.com/streamnative/oxia/proto"
+)
+
+type notificationDispatcher struct {
+	lc     *leaderController
+	id     int64
+	req    *proto.NotificationsRequest
+	stream proto.OxiaClient_GetNotificationsServer
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	closeCh chan any
+	log     *slog.Logger
+}
+
+var notificationDispatcherIdGen atomic.Int64
+
+func startNotificationDispatcher(lc *leaderController, req *proto.NotificationsRequest, stream proto.OxiaClient_GetNotificationsServer) error {
+	nd := &notificationDispatcher{
+		lc:      lc,
+		id:      notificationDispatcherIdGen.Add(1),
+		req:     req,
+		stream:  stream,
+		log:     lc.log.With(slog.String("component", "notification-dispatcher")),
+		closeCh: make(chan any),
+	}
+
+	lc.Lock()
+	lc.notificationDispatchers[nd.id] = nd
+	lc.Unlock()
+
+	// Create a context for handling this stream
+	nd.ctx, nd.cancel = context.WithCancel(stream.Context())
+
+	go common.DoWithLabels(
+		nd.ctx,
+		map[string]string{
+			"oxia":  "dispatch-notifications",
+			"shard": fmt.Sprintf("%d", lc.shardId),
+			"peer":  common.GetPeer(stream.Context()),
+		},
+		func() {
+			if err := nd.dispatchNotifications(); err != nil && !errors.Is(err, context.Canceled) {
+				nd.log.Warn(
+					"Failed to dispatch notifications",
+					slog.Any("error", err),
+					slog.String("peer", common.GetPeer(stream.Context())),
+				)
+				nd.cancel()
+			}
+
+			close(nd.closeCh)
+
+			// Clean up dispatcher for leader controller map
+			nd.lc.Lock()
+			delete(nd.lc.notificationDispatchers, nd.id)
+			nd.lc.Unlock()
+		},
+	)
+
+	select {
+	case <-lc.ctx.Done():
+		// Leader is getting closed
+		nd.cancel()
+		return lc.ctx.Err()
+
+	case <-nd.ctx.Done():
+		return nd.ctx.Err()
+
+	case <-stream.Context().Done():
+		// The stream is getting closed
+		nd.cancel()
+		return stream.Context().Err()
+	}
+}
+
+func (nd *notificationDispatcher) dispatchNotifications() error {
+	nd.log.Debug(
+		"Dispatch notifications",
+		slog.Any("start-offset-exclusive", nd.req.StartOffsetExclusive),
+	)
+
+	var offsetInclusive int64
+	if nd.req.StartOffsetExclusive != nil {
+		offsetInclusive = *nd.req.StartOffsetExclusive + 1
+	} else {
+		nd.lc.Lock()
+		qat := nd.lc.quorumAckTracker
+		nd.lc.Unlock()
+
+		if qat == nil {
+			return errors.New("leader is not yet ready")
+		}
+		commitOffset := qat.CommitOffset()
+
+		// The client is creating a new notification stream and wants to receive the notification from the next
+		// entry that will be written.
+		// In order to ensure the client will positioned on a given offset, we need to send a first "dummy"
+		// notification. The client will wait for this first notification before making the notification
+		// channel available to the application
+		nd.log.Debug(
+			"Sending first dummy notification",
+			slog.Int64("commit-offset", commitOffset),
+		)
+		if err := nd.stream.Send(&proto.NotificationBatch{
+			Shard:         nd.lc.shardId,
+			Offset:        commitOffset,
+			Timestamp:     0,
+			Notifications: nil,
+		}); err != nil {
+			return err
+		}
+
+		offsetInclusive = commitOffset + 1
+	}
+
+	return nd.iterateOverNotifications(offsetInclusive)
+}
+
+func (nd *notificationDispatcher) iterateOverNotifications(startOffsetInclusive int64) error {
+	lc := nd.lc
+	offsetInclusive := startOffsetInclusive
+	for nd.ctx.Err() == nil {
+		notifications, err := lc.db.ReadNextNotifications(nd.ctx, offsetInclusive)
+		if err != nil {
+			return err
+		}
+
+		nd.log.Debug(
+			"Got a new list of notification batches",
+			slog.Int("list-size", len(notifications)),
+		)
+
+		for _, n := range notifications {
+			if err := nd.stream.Send(n); err != nil {
+				return err
+			}
+		}
+
+		offsetInclusive += int64(len(notifications))
+	}
+
+	return nd.ctx.Err()
+}
+
+func (nd *notificationDispatcher) close() {
+	// Wait for dispatcher stream to be fully closed
+	<-nd.closeCh
+}

